base_layer/core/src/base_node/state_machine.rs

```diff
@@ -140,31 +134,36 @@ impl<B: BlockchainBackend + 'static> BaseNodeStateMachine<B> {
         let mut state = Starting(states::Starting);
         let mut shared_state = self;
         loop {
-            if shared_state.is_stop_requested() {
-                break;
-            }
+            let interupt_signal = shared_state.get_interrupt_signal();
             let _ = shared_state.event_sender.send(state.clone()).await;
             let next_event = match &mut state {
-                Starting(s) => s.next_event(&shared_state).await,
-                BlockSync(s, network_tip, sync_peers) => s.next_event(&mut shared_state, network_tip, sync_peers).await,
-                Listening(s) => s.next_event(&mut shared_state).await,
+                Starting(s) => interrupt_state(interupt_signal, s.next_event(&shared_state)).await,
+                BlockSync(s, network_tip, sync_peers) => {
+                    interrupt_state(
+                        interupt_signal,
+                        s.next_event(&mut shared_state, network_tip, sync_peers),
+                    )
+                    .await
+                },
+                Listening(s) => interrupt_state(interupt_signal, s.next_event(&mut shared_state)).await,
                 Shutdown(_) => break,
             };
+
             debug!(
                 target: LOG_TARGET,
                 "=== Base Node event in State [{}]:  {:?}", state, next_event
             );
             state = BaseNodeStateMachine::<B>::transition(state, next_event);
         }
     }
+}
 
-    /// Checks the value of the interrupt flag and returns a `FatalError` event if the flag is true. Otherwise it
-    /// returns the `default` event.
-    pub fn check_interrupt(flag: &AtomicBool, default: StateEvent) -> StateEvent {
-        if flag.load(Ordering::SeqCst) {
-            StateEvent::FatalError("User interrupted".into())
-        } else {
-            default
-        }
+async fn interrupt_state<F>(signal: ShutdownSignal, fut: F) -> StateEvent
+where F: Future<Output = StateEvent> {
+    futures::pin_mut!(fut);
+    // If future A and B are both ready `future::select` will prefer A
+    match future::select(signal, fut).await {
+        Either::Left(_) => StateEvent::UserQuit,
+        Either::Right((state, _)) => state,
```

Comment:
Can this be tidied up? The LHS is a very pristine, self-documenting state machine; and it would be a pity to lose it

maybe the shutdown signal can be select!ed for one level up?

Reply:
Agree that is muddies the implementation - I wanted to ensure that the `(_, Shutdown)` state transition still runs (although it does nothing at the moment, it may cause bugs in future if the transition doesn't run) - I'll see how to neaten up those branches

Reply:
`(_, UserQuit)` *